File: backend/app/schemas/sharia.py
"""Sharia-compliance verdict types (CR069 Phase 1).

The `halal` mandate flag is enforced by a *sourced allowlist* — the published
constituents of the S&P 500 Sharia Industry Exclusions Index (AAOIFI standard,
screened by S&P Dow Jones), read from the SPUS ETF's daily-transparency holdings
CSV. It is NOT a computed ratio screen: `app.trading_math.screening.sharia_screen`
stays dormant (CR069 constraint 4) until a source supplies real debt / liquid-asset
/ impermissible-income figures.

Three states, never two (CR069 constraint 2 / DEF084):
  PASS         — in the compliant set. Tradeable.
  SCREENED_OUT — in the parent index (S&P 500) but absent from the compliant set.
                 A real exclusion — blocked.
  UNKNOWN      — not in the parent index at all. The standard never looked at it,
                 so it is *no ruling either way*. Permitted, with the disclosure
                 attached (G3 RESOLVED 2026-07-23: permit + disclose).
  UNAVAILABLE  — the source could not be refreshed / is stale beyond its window.
                 The flag pauses loudly (CR069 constraint 3 / CR040) — never a
                 silent fall back to a stale set.

These are value types only; the fetcher, cache and resolver live in
`app.services.sharia_universe` (network belongs in the service layer).
"""

from __future__ import annotations

from datetime import date
from enum import Enum

from pydantic import BaseModel


class ShariaStatus(str, Enum):
    PASS = "pass"
    SCREENED_OUT = "screened_out"
    UNKNOWN = "unknown"
    UNAVAILABLE = "unavailable"


def _as_of_str(as_of: date | None) -> str:
    return as_of.isoformat() if as_of is not None else "unknown"


class ShariaVerdict(BaseModel):
    """A per-ticker Sharia screen verdict carrying its provenance.

    The verdict is NOT a bare boolean by design (CR069 §Acceptance 4): the
    standard, source and as-of date travel with it so every surface that renders
    a trade outcome — and every agent prompt — can name what screened the ticker
    and when. A permitted-unknown trade that says nothing is a silent pass on an
    observance decision, which is this CR's failure class in the other direction.
    """

    status: ShariaStatus
    ticker: str
    standard: str  # e.g. "AAOIFI"
    source: str  # e.g. "S&P 500 Sharia Industry Exclusions Index (via SPUS)"
    as_of: date | None = None

    @property
    def is_blocking(self) -> bool:
        """SCREENED_OUT and UNAVAILABLE block a trade; PASS and UNKNOWN permit it.

        UNKNOWN must NOT block (G3) — an unknown ticker is *no ruling*, not a soft
        no. If UNKNOWN ever reaches a violations list the ruling is inverted.
        """
        return self.status in (ShariaStatus.SCREENED_OUT, ShariaStatus.UNAVAILABLE)

    def message(self) -> str:
        """User-facing line, AMI by name (never "the AI"). Names the standard,
        source and as-of date wherever a verdict is shown (CR069 constraint 1)."""
        t = self.ticker.upper()
        when = _as_of_str(self.as_of)
        if self.status is ShariaStatus.PASS:
            return f"{t} passes the {self.standard} screen ({self.source}, as of {when})."
        if self.status is ShariaStatus.SCREENED_OUT:
            return (
                f"{t} is in the parent index but does not pass the {self.standard} "
                f"screen ({self.source}, as of {when}), so this mandate won't trade it."
            )
        if self.status is ShariaStatus.UNKNOWN:
            return (
                f"{t} isn't in the parent index, so the {self.standard} screen AMI "
                f"uses hasn't reviewed it. That's not a ruling either way — AMI doesn't know."
            )
        # UNAVAILABLE
        return (
            f"AMI couldn't refresh the {self.standard} Sharia screen (last updated "
            f"{when}). The halal filter is paused until it can."
        )

File: mobile/lib/models/sharia.dart
/// Mobile mirror of the backend's Sharia verdict type (CR069 Phase 1b).
///
/// Hand-mirrored from `backend/app/schemas/sharia.py` — the backend↔mobile seam
/// is over-the-wire JSON with no codegen, so this file is the *only* thing
/// keeping the two sides in step. A rename on the backend fails silently here
/// behind a `??` default, which is the failure mode DEF084 and CR040 both point
/// at. `ShariaStatus.fromWire` therefore returns null on an unrecognised value
/// rather than defaulting to a state, so a drifted enum surfaces as "no verdict"
/// instead of a confidently wrong one.
///
/// Four states, never two (CR069 constraint 2):
///   pass         — in the compliant set. Tradeable.
///   screenedOut  — in the parent index (S&P 500) but absent from the compliant
///                  set. A real exclusion — blocks the trade.
///   unknown      — not in the parent index at all. The standard never looked at
///                  it, so it is *no ruling either way*. PERMITTED, with the
///                  disclosure attached (G3, resolved 2026-07-23). Its copy
///                  therefore lands on a SUCCESSFUL trade, not on a rejection,
///                  and must never read as a warning that the trade was risky.
///   unavailable  — the source could not be refreshed / is stale beyond its
///                  window. The flag pauses loudly (CR040) — never a silent
///                  stale read.
///
/// The strings are NOT built here. The backend composes an English sentence in
/// `ShariaVerdict.message()`; the app renders its own localized copy from the
/// ARB so AR/MS can diverge. This type carries only the structured facts —
/// status, ticker, standard, source, as-of date — which are what constraint 1
/// requires travel with every verdict.
library;

enum ShariaStatus {
  pass,
  screenedOut,
  unknown,
  unavailable;

  /// Wire values as emitted by `backend/app/schemas/sharia.py::ShariaStatus`.
  static const Map<String, ShariaStatus> _wire = {
    'pass': ShariaStatus.pass,
    'screened_out': ShariaStatus.screenedOut,
    'unknown': ShariaStatus.unknown,
    'unavailable': ShariaStatus.unavailable,
  };

  /// Null on an unrecognised wire value — see the library doc. Do NOT add a
  /// default; a silent fallback re-creates the exact class of bug this type
  /// exists to close.
  static ShariaStatus? fromWire(Object? raw) =>
      raw is String ? _wire[raw] : null;

  /// SCREENED_OUT and UNAVAILABLE block a trade; PASS and UNKNOWN permit it.
  /// Mirrors `ShariaVerdict.is_blocking`. UNKNOWN must never block (G3) — an
  /// unknown ticker is *no ruling*, not a soft no.
  bool get isBlocking =>
      this == ShariaStatus.screenedOut || this == ShariaStatus.unavailable;
}

class ShariaVerdict {
  const ShariaVerdict({
    required this.status,
    required this.ticker,
    required this.standard,
    required this.source,
    this.asOf,
  });

  final ShariaStatus status;
  final String ticker;

  /// e.g. "AAOIFI". Named wherever the verdict is shown (constraint 1).
  final String standard;

  /// e.g. "S&P 500 Sharia Industry Exclusions Index (via SPUS)".
  final String source;

  /// ISO date from the source's own as-of stamp. Null only when the backend
  /// could not stamp one (an unavailable/paused screen).
  final DateTime? asOf;

  bool get isBlocking => status.isBlocking;

  /// Returns null when the payload carries no verdict at all (halal flag off,
  /// or a backend that does not yet serialize the field) and when the status
  /// is unrecognised. A caller that gets null must render NOTHING rather than
  /// guess a state.
  static ShariaVerdict? fromJson(Map<String, dynamic>? j) {
    if (j == null) return null;
    final status = ShariaStatus.fromWire(j['status']);
    if (status == null) return null;
    return ShariaVerdict(
      status: status,
      ticker: (j['ticker'] as String? ?? '').toUpperCase(),
      standard: j['standard'] as String? ?? '',
      source: j['source'] as String? ?? '',
      asOf: _parseDate(j['as_of']),
    );
  }

  static DateTime? _parseDate(Object? raw) =>
      raw is String ? DateTime.tryParse(raw) : null;
}
